galaxy/api/download/tests/test_download_views.py

```diff
@@ -0,0 +1,128 @@
+#
+# You should have received a copy of the Apache License
+# along with Galaxy.  If not, see <http://www.apache.org/licenses/>.
+
+import importlib
+from urllib import parse as urlparse
+from unittest import mock
+
+from django.core.files.uploadedfile import SimpleUploadedFile
+from django.contrib.auth import get_user_model
+from django.test import override_settings
+from pulpcore.app import models as pulp_models
+from rest_framework import status as http_codes
+from rest_framework.test import APITestCase
+
+from galaxy.main import models
+
+
+UserModel = get_user_model()
+
+
+class TestArtifactDownloadView(APITestCase):
+    download_url = '/download/{filename}'
+    mazer_user_agent = 'Mazer/0.4.0 (linux; python:3.6.8) ansible_galaxy/0.4.0'
+
+    @classmethod
+    def setUpClass(cls):
+        # Parameters of `storages.backends.s3boto3.S3Boto3Storage` class are
+        # initialized as class-attributes on module load.
+        # Because `storages.backends.s3boto3` module can be imported even
+        # before S3 specific test is executed overriding parameters locally
+        # is not effective.
+        # Thus we import `storages.backends.s3boto3` module before any
+        # request to download view is called.
+        super().setUpClass()
+        settings = dict(
+            AWS_ACCESS_KEY_ID='key',
+            AWS_SECRET_ACCESS_KEY='secret',
+            AWS_STORAGE_BUCKET_NAME='test',
+            AWS_DEFAULT_ACL=None,
+        )
+        with override_settings(**settings):
+            importlib.import_module('storages.backends.s3boto3')
+
+    def setUp(self):
+        super().setUp()
+        self.sha256 = \
+            '01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b'
+        self.storage_path = f'artifact/{self.sha256[:2]}/{self.sha256[2:]}'
+        self.filename = 'mynamespace-mycollection-1.2.3.tar.gz'
+
+        self.user = UserModel.objects.create_user(
+            username='testuser', password='secret')
+        self.namespace = models.Namespace.objects.create(name='mynamespace')
+        self.namespace.owners.set([self.user])
+
+        self.collection = models.Collection.objects.create(
+            namespace=self.namespace, name='mycollection')
+        self.version = models.CollectionVersion.objects.create(
+            collection=self.collection, version='1.2.3')
+
+    def _create_artifact(self):
+        self.artifact = pulp_models.Artifact.objects.create(
+            size=427611,
+            sha256=self.sha256,
+            file=SimpleUploadedFile(self.storage_path, ''),
+        )
+        self.ca = pulp_models.ContentArtifact.objects.create(
+            content=self.version, artifact=self.artifact,
+            relative_path=self.filename
+        )
+
+    @override_settings(
+        MEDIA_ROOT='',
+        DEFAULT_FILE_STORAGE='storages.backends.s3boto3.S3Boto3Storage',
+        AWS_DEFAULT_ACL=None,
+    )
+    @mock.patch('storages.backends.s3boto3.S3Boto3Storage.save')
```

Comment:
Where do the AWS credentials come in and is there a guarantee the tests don't run with actual S3 access in case something outside the patch tries to do something?

Reply:
AWS credentials are set in `setUpClass` method, due to how settings are initialized in `S3Boto3Storage` class (see my comment).

Actual S3 credentials are never stored in repository and test environment has no access to them, so it's impossible unless somebody is running tests in production environment with production config.